File: lib/src/test/java/io/xj/lib/meme/MemeTaxonomyTest.java
package io.xj.lib.meme;

import com.google.common.collect.ImmutableMap;
import org.junit.Test;

import java.util.List;

import static org.junit.Assert.*;

/**
 TemplateConfig has Meme categories
 https://www.pivotaltracker.com/story/show/181801646
 <p>
 A template configuration has a field called `memeTaxonomy` which defines the taxonomy of memes.
 <p>
 For example, this might look like
 <p>
 ```
 memeTaxonomy=CITY[CHICAGO,DENVER,PHILADELPHIA]
 ```
 <p>
 That would tell XJ about the existence of a meme category called City with values `CHICAGO`, `DENVER`, and `PHILADELPHIA`. And these would function as exclusion like numeric memes, e.g. after content having `CHICAGO` is chosen, we can choose nothing with `DENVER` or `PHILADELPHIA`.
 */
public class MemeTaxonomyTest {
  @Test
  public void test_fromString_toString() {
    var subject = MemeTaxonomy.fromString("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]");

    assertEquals("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]", subject.toString());
  }

  @Test
  public void test_fromListOfMaps_toListOfMaps() {
    var subject = MemeTaxonomy.fromList(List.of(
      ImmutableMap.of(
        "name", "CO111LOR",
        "memes", List.of("RED ", "GR333EEN", "BLUE#@")
      ),
      ImmutableMap.of(
        "name", "SIZ$%@E",
        "memes", List.of("L44A&*(RGE", "MED$IUM", "SMA)(&&LL")
      )));

    assertEquals(List.of(
      ImmutableMap.of(
        "name", "COLOR",
        "memes", List.of("RED", "GREEN", "BLUE")
      ),
      ImmutableMap.of(
        "name", "SIZE",
        "memes", List.of("LARGE", "MEDIUM", "SMALL")
      )
    ), subject.toList());
  }

  @Test
  public void testStripNonAlphabetical() {
    var subject = MemeTaxonomy.fromString("COLOR [RED, GREEN, BLUE];    SIZE [LARGE, MEDIUM, SMALL ]");

    assertEquals("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]", subject.toString());
  }

  @Test
  public void getCategories() {
    var subject = MemeTaxonomy.fromString("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]");

    assertEquals(2, subject.getCategories().size());
    assertEquals("COLOR", subject.getCategories().get(0).getName());
    assertEquals("RED", subject.getCategories().get(0).getMemes().get(0));
    assertEquals("GREEN", subject.getCategories().get(0).getMemes().get(1));
    assertEquals("BLUE", subject.getCategories().get(0).getMemes().get(2));
    assertEquals("SIZE", subject.getCategories().get(1).getName());
    assertEquals("LARGE", subject.getCategories().get(1).getMemes().get(0));
    assertEquals("MEDIUM", subject.getCategories().get(1).getMemes().get(1));
    assertEquals("SMALL", subject.getCategories().get(1).getMemes().get(2));
  }

  @Test
  public void isAllowed() {
    assertTrue(MemeTaxonomy.fromString("CITY[CHICAGO,DENVER,PHILADELPHIA]").isAllowed(List.of("PEACHES")));
    assertTrue(MemeTaxonomy.fromString("CITY[CHICAGO,DENVER,PHILADELPHIA]").isAllowed(List.of("DENVER")));
    assertFalse(MemeTaxonomy.fromString("CITY[CHICAGO,DENVER,PHILADELPHIA]").isAllowed(List.of("DENVER", "PHILADELPHIA")));
    assertTrue(MemeTaxonomy.fromString("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed(List.of("RED", "LARGE")));
    assertTrue(MemeTaxonomy.fromString("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed(List.of("GREEN", "MEDIUM", "PEACHES")));
    assertFalse(MemeTaxonomy.fromString("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed(List.of("RED", "BLUE", "LARGE")));
    assertFalse(MemeTaxonomy.fromString("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed(List.of("GREEN", "MEDIUM", "PEACHES", "SMALL")));
  }

}

File: engine/test/entities/meme/MemeTaxonomyTest.cpp
// Copyright (c) 1999-2022, XJ Music Inc. (https://xj.io) All Rights Reserved.

#include <gtest/gtest.h>
#include <set>

#include "xjmusic/entities/meme/MemeTaxonomy.h"

using namespace XJ;


/**
 TemplateConfig has Meme categories
 https://github.com/xjmusic/xjmusic/issues/209
 <p>
 A template configuration has a field called `memeTaxonomy` which defines the taxonomy of memes.
 <p>
 For example, this might look like
 <p>
 ```
 memeTaxonomy=CITY[CHICAGO,DENVER,PHILADELPHIA]
 ```
 <p>
 That would tell XJ about the existence of a meme category called City with values `CHICAGO`, `DENVER`, and `PHILADELPHIA`. And these would function as exclusion like numeric memes, e.g. after content having `CHICAGO` is chosen, we can choose nothing with `DENVER` or `PHILADELPHIA`.
 */

TEST(MemeTaxonomyTest, DefaultCategoryName) {
  std::set<MapStringToOneOrManyString> input = {
      {{"memes", std::set<std::string>{"WINTER", "SPRING", "SUMMER", "FALL"}}},
  };

  auto subject = MemeTaxonomy::fromSet(input);

  EXPECT_EQ("CATEGORY", subject.getCategories().begin()->getName());
}

TEST(MemeTaxonomyTest, CategoryHasMemes) {
  std::set<MapStringToOneOrManyString> input = {
      {{"memes", std::set<std::string>{"WINTER", "SPRING", "SUMMER", "FALL"}}},
  };

  MemeTaxonomy subject(input);

  ASSERT_TRUE(subject.getCategories().begin()->hasMemes());
}

TEST(MemeTaxonomyTest, FromStringToString) {
  MemeTaxonomy subject("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]");

  ASSERT_EQ("COLOR[BLUE,GREEN,RED];SIZE[LARGE,MEDIUM,SMALL]", subject.toString());
}

TEST(MemeTaxonomyTest, FromSetToList) {
  std::set<MapStringToOneOrManyString> input = {
      {{"name", "CO111LOR"}, {"memes", std::set<std::string>{"RED ", "GR333EEN", "BLUE#@"}}},
      {{"name", "SIZ$%@E"}, {"memes", std::set<std::string>{"L44A&*(RGE", "MED$IUM", "SMA)(&&LL"}}}
  };

  MemeTaxonomy subject(input);

  std::set<MapStringToOneOrManyString> expected = {
      {{"name", "COLOR"}, {"memes", std::set<std::string>{"RED", "GREEN", "BLUE"}}},
      {{"name", "SIZE"}, {"memes", std::set<std::string>{"LARGE", "MEDIUM", "SMALL"}}}
  };

  ASSERT_EQ(expected, subject.toList());
}

TEST(MemeTaxonomyTest, TestStripNonAlphabetical) {
  MemeTaxonomy subject("COLOR [RED, GREEN, BLUE];    SIZE [LARGE, MEDIUM, SMALL ]");

  ASSERT_EQ("COLOR[BLUE,GREEN,RED];SIZE[LARGE,MEDIUM,SMALL]", subject.toString());
}

TEST(MemeTaxonomyTest, GetCategories) {
  MemeTaxonomy subject("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]");

  std::vector<MemeCategory> categories;
  for (const auto &category: subject.getCategories()) {
    categories.push_back(category);
  }
  std::sort(categories.begin(), categories.end());

  ASSERT_EQ(2, categories.size());
  ASSERT_EQ("COLOR", categories[0].getName());
  ASSERT_TRUE(categories[0].getMemes().find("RED") != categories[0].getMemes().end());
  ASSERT_TRUE(categories[0].getMemes().find("GREEN") != categories[0].getMemes().end());
  ASSERT_TRUE(categories[0].getMemes().find("BLUE") != categories[0].getMemes().end());
  ASSERT_EQ("SIZE", categories[1].getName());
  ASSERT_TRUE(categories[1].getMemes().find("LARGE") != categories[1].getMemes().end());
  ASSERT_TRUE(categories[1].getMemes().find("MEDIUM") != categories[1].getMemes().end());
  ASSERT_TRUE(categories[1].getMemes().find("SMALL") != categories[1].getMemes().end());
}

TEST(MemeTaxonomyTest, IsAllowed) {
  ASSERT_TRUE(MemeTaxonomy("CITY[CHICAGO,DENVER,PHILADELPHIA]").isAllowed({"PEACHES"}));
  ASSERT_TRUE(MemeTaxonomy("CITY[CHICAGO,DENVER,PHILADELPHIA]").isAllowed({"DENVER"}));
  ASSERT_FALSE(MemeTaxonomy("CITY[CHICAGO,DENVER,PHILADELPHIA]").isAllowed({"DENVER", "PHILADELPHIA"}));
  ASSERT_TRUE(MemeTaxonomy("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed({"RED", "LARGE"}));
  ASSERT_TRUE(MemeTaxonomy("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed({"GREEN", "MEDIUM", "PEACHES"}));
  ASSERT_FALSE(MemeTaxonomy("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed({"RED", "BLUE", "LARGE"}));
  ASSERT_FALSE(MemeTaxonomy("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed({"GREEN", "MEDIUM", "PEACHES", "SMALL"}));
}

TEST(MemeTaxonomyTest, IsAllowed_alreadyPresentFromTaxonomy) {
  ASSERT_TRUE(MemeTaxonomy("CITY[ABERDEEN,NAGOYA]").isAllowed({"ABERDEEN", "ABERDEEN"}));
}

